### PR TITLE
feat: add remove mesh by blend shape component

### LINKED ISSUE #43: remove mesh based on blendshape

### PR DESCRIPTION
**Pull request by nil-vr**
Added by anatawa12: Closes #43

This PR adds a new component for removing mesh by blend shapes (#43). It's based on the "Remove Mesh in Box" and "Freeze BlendShapes" components. If one of the checked blend shapes translates all of the vertices of a triangle by more than "Tolerance" then the triangle is removed from the mesh.

It might also make sense to remove these blend shapes. However, they are still used by triangles that are only partially affected by the blend shapes. Maybe `FreezeBlendShapesProcessor` could be called at the end of `RemoveMeshByBlendShapeProcessor`, but `FreezeBlendShapesProcessor` uses `FreezeBlendShapes` and `RemoveMeshByBlendShapeProcessor` doesn't have one.

### CODE REVIEW

**Review by anatawa12 — CHANGES_REQUESTED** [review]
Thank you for implementing RemoveMeshByBlendShape.

> If one of the checked blend shapes translates all of the vertices of a triangle by more than "Tolerance" then the triangle is removed from the mesh.

I think removing triangle if **any** of the vertices is translated.
In RemoveMeshInBox, it's intended to remove polygon in box.
On the other hand, I think RemoveMeshByBlendShape should remove transformed vertices.

In addition, can you add to changelog?

File: Runtime/RemoveMeshByBlendShape.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Anatawa12.AvatarOptimizer
+{
+    [AddComponentMenu("Avatar Optimizer/Remove Mesh By Blend Shape")]
+    [RequireComponent(typeof(SkinnedMeshRenderer))]
+    [DisallowMultipleComponent]
+    internal class RemoveMeshByBlendShape : EditSkinnedMeshComponent
+    {
+        public PrefabSafeSet.StringSet shapeKeysSet;
+        [CL4EELocalized("RemoveMeshByBlendShape:prop:Tolerance")]
```
[the remark this answers is not in the document]
Comment (nil-vr) [reply Runtime/RemoveMeshByBlendShape.cs:13]: I can do it for English, but my Japanese description probably would not be very good.

[the remark this answers is not in the document]
Comment (anatawa12) [reply Runtime/RemoveMeshByBlendShape.cs:13]: we will add japanese translation later so no problem.

File: Runtime/RemoveMeshByBlendShape.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        public double Tolerance = 0.001;
```
Comment (anatawa12) [inline Runtime/RemoveMeshByBlendShape.cs:14]: Pease use lowerCamelCase for unity serialized properties.
In other words, please rename to `tolerance`.

**Comment by nil-vr** [comment]
This still doesn't remove the blend shapes that are removed, so a separate freeze blend shapes behavior should be used.

Having used it on some avatars, I wonder if this is the right behavior and editor. Maybe it'd be better if deleting and freezing were in the same behavior. Even if this behavior were to remove the blend shapes, they would still appear in the freeze blend shapes editor.

**Comment by anatawa12** [comment]
> Maybe it'd be better if deleting and freezing were in the same behavior.

I think we should remove the blendshapes in `RemoveMeshByBlendShape` component.
I think we can pull some of `FreezeBlendShapesProcessor` into an static method and use the static method is one good way to implement.

> Even if this behavior were to remove the blend shapes, they would still appear in the freeze blend shapes editor.

In SkinnedMeshRenderer, It's expected because AAO is non-desctructive avatar modification tool. If you mean you can find in other AAO components, we have to remove blendshape name in `MeshInfoComputer` to tell other AAO components that it's removed.

**Comment by nil-vr** [comment]
`RemoveMeshByBlendShapeProcessor` now calls a static method from `FreezeBlendShapeProcessor` to remove the blend shapes without needing to also select the same blend shapes in a separate `FreezeBlendShapesComponent`.

It isn't smart enough to handle the case where another blend shape transforms a subset of the vertices transformed by the deleted blend shapes. For example, Mamehinata has Shoulder_OFF and Shoulder_OFF_left. If you delete the vertices affected by Shoulder_OFF, Shoulder_OFF is deleted from the skinned mesh renderer, but a useless Shoulder_OFF_left remains. Maybe there could be a "Remove unused blend shapes" feature that would clean everything up using similar logic to find blend shapes that transform no vertex by more than some tolerance value.

However, for my workflow it doesn't matter. I freeze every blend shape on that mesh anyway.

**Review by anatawa12 — APPROVED** [review]
It looks good for me. Thanks a lot!

I think freezing meaningless BlendShapes is out of scope of this component. we can add some pass in Automatic Configuration component to remove such a BlendShapes in the feature.

**Pull request merged**